Include/Rover/Model.hpp

```diff
@@ -0,0 +1,190 @@
+#ifndef ROVER_MODEL_HPP
+#define ROVER_MODEL_HPP
+#include <algorithm>
+#include <numeric>
+#include <random>
+#include <type_traits>
+#include <vector>
+#include "ScalarView.hpp"
+
+namespace {
+  template<typename R, typename X, typename Y>
+  constexpr std::enable_if_t<!std::is_same_v<std::decay_t<R>,
+      std::decay_t<X>> || !std::is_same_v<std::decay_t<R>, std::decay_t<Y>>,
+      bool> solve_basis(R&, X&, Y&) {
+    return false;
+  }
+
+  template<typename R, typename X, typename Y>
+  std::enable_if_t<std::is_same_v<std::decay_t<R>, std::decay_t<X>> &&
+      std::is_same_v<std::decay_t<R>, std::decay_t<Y>>, bool>
+      solve_basis(R& r, X& x, Y& y) {
+    if(x == y) {
+      return false;
+    }
+    if(x + y == x) {
+      r = x;
+    }
+    else {
+      r = y;
+    }
+    return true;
+  }
+
+  template<typename C, typename V, typename B>
+  constexpr std::enable_if_t<!std::is_same_v<V, B>> apply_basis(C&, const V&,
+    const B&) {}
+
+  template<typename C, typename V, typename B>
+  constexpr std::enable_if_t<std::is_same_v<V, B>> apply_basis(C& c,
+      const V& v, const B& b) {
+    c = v / b;
+  }
+}
+
+namespace Rover { 
+
+  //! Single-value regression (prediction) model.
+  /*!
+    \tparam A The type of the regression algorithm.
+    \tparam T The type of the trial.
+  */
+  template<typename A, typename T>
+  class Model {
+    public:
+
+      //! The type of the regression algorithm.
+      using Algorithm = A;
+
+      //! The computation type used by the algorithm.
+      using CompType = typename Algorithm::CompType;
```

Comment:
Don't introduce new abbreviations. You are welcome to use existing abbreviations that are universal. In this case you can rename ```CompType``` to ```Type``` or ```ComputationType```.